resources/src/state/types.ts

```diff
@@ -238,24 +270,51 @@ export interface GraphGetResponse {
     roots: string[];
     hasCycle: boolean;
   };
-  meta: {
-    schemaVersion: number;
-    timestamp: string;
-    from_cache: boolean;
-  };
 }
 
 /**
  * Response from labkiHierarchyGet API.
  */
-export interface HierarchyGetResponse {
+export interface HierarchyGetResponse extends ActionAPIResponseBase {
   repo_url: string;
   ref: string;
   hash: string;
   hierarchy: Hierarchy;
-  meta: {
-    schemaVersion: number;
-    timestamp: string;
-    from_cache: boolean;
-  };
+}
+
+export interface OperationsStatusRequest extends ActionAPIRequestBase<'labkiOperationsStatus'> {
+  operation_id: number;
+}
+
+export interface OperationsStatusResponse extends ActionAPIResponseBase {
+  status: string;
+  operation_id: string;
+  message: string;
+}
+
+// Type unions for api.get and api.post methods
+// using a combination of conditional types with a generic and a mapping
+// so that the `action` within the request can be linked to the response type
+
+export type ActionAPIRequest<T extends ActionAPIName> = T extends 'labkiReposList'
+  ? ActionAPIRequestBase<T>
+  : T extends 'labkiReposAdd'
+    ? ReposAddRequest
+    : T extends 'labkiGraphGet'
+      ? RepoRefRequest<T>
+      : T extends 'labkiHierarchyGet'
+        ? RepoRefRequest<T>
+        : T extends 'labkiPacksAction'
+          ? PacksActionRequest
+          : T extends 'labkiOperationsStatus'
+            ? OperationsStatusRequest
+            : never;
+
+export interface ActionAPIResponseMap {
+  labkiReposList: ReposListResponse;
+  labkiReposAdd: ReposAddResponse;
+  labkiGraphGet: GraphGetResponse;
+  labkiHierarchyGet: HierarchyGetResponse;
+  labkiPacksAction: PacksActionWrapper;
+  labkiOperationsStatus: OperationsStatusResponse;
```

Comment:
this little bit of magic along with the `mediawiki.d.ts` changes make it so typescript can infer the structure of the request and response objects from the value of the `action` parameter in the request object.